sisyphus/simple_linux_utility_for_resource_management_engine.py

```diff
@@ -358,7 +367,7 @@ def task_state(self, task, task_id):
         elif state in ["PENDING", "CONFIGURING"]:
             return STATE_QUEUE
         else:
-            return STATE_UNKNOWN
+            return STATE_QUEUE_ERROR
```

Comment:
No, that's wrong. This else is the case where a task is legitimately not found in the queue (i.e. finished or not yet submitted). The queue error would be in the ` qs == []` case. 

Although I don't think we can arrive there any more with the addition of `if retval != 0: continue`

Reply:
> The queue error would be in the ` qs == []` case.

See my other comment. We should not check for `qs == []`. We should use a `try: ... except CalledProcessError: return STATE_QUEUE_ERROR`.

I think the `return STATE_QUEUE_ERROR` here is fine. We only get here if we have an unexpected state. That doesn't mean that the engine does not know about the job. It knows about the job but fails to parse the state.

Reply:
> I think the `return STATE_QUEUE_ERROR` here is fine. We only get here if we have an unexpected state. That doesn't mean that the engine does not know about the job. It knows about the job but fails to parse the state.

I beg to differ. I think we arrive here when the job has not yet been submitted and with this change we cannot submit any job.

Reply:
But you have this code above:
```python
queue_state = self.queue_state()
qs = queue_state[task_name]
```
When `qs` is empty, that means that there are no corresponding tasks reported by `squeue` (and thus it should return `STATE_UNKNOWN`, which means exactly that: as you said, "this job is not known to the engine").
When `qs` is non-empty, it means there *are* corresponding tasks reported by `squeue`. So that by your definition implies that it cannot return `STATE_UNKNOWN`, because it is known to the engine. If the state is not one of the expected states, it means the task is known but something unexpected happened, i.e. that's exactly what `STATE_QUEUE_ERROR` means.

Reply:
Yes, sorry, I had assumed the empty list would result from a faulty `squeue` call. But in fact it is just what we get from the defaultdict if a key is missing. 

I take back the original comment